certora/specs/ERC20.spec

```diff
@@ -1,28 +1,32 @@
-import "helpers/helpers.spec"
-import "methods/IERC20.spec"
-import "methods/IERC2612.spec"
+import "helpers/helpers.spec";
+import "methods/IERC20.spec";
+import "methods/IERC2612.spec";
 
 methods {
     // non standard ERC20 functions
-    increaseAllowance(address,uint256) returns (bool)
-    decreaseAllowance(address,uint256) returns (bool)
+    function increaseAllowance(address,uint256) external returns (bool);
+    function decreaseAllowance(address,uint256) external returns (bool);
 
     // exposed for FV
-    mint(address,uint256)
-    burn(address,uint256)
+    function mint(address,uint256) external;
+    function burn(address,uint256) external;
 }
 
 /*
 ┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
 │ Ghost & hooks: sum of all balances                                                                                  │
 └─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
 */
-ghost sumOfBalances() returns uint256 {
-  init_state axiom sumOfBalances() == 0;
+ghost mathint sumOfBalances {
+    init_state axiom sumOfBalances == 0;
+}
+
+hook Sload uint256 balance _balances[KEY address addr] STORAGE {
```

Comment:
I added a comment explaining why it's needed. However, I don't think removing it should be ideal. 

The reason why we need it now seems to be because the casting in CVL2 works differently and previously (iiuc) it avoided any operation overflowing the `uint256` as if we were doing a `require_uint256()` cast. Something I discussed with @frangio is that this change may have uncovered an undesired behavior in this spec.

I couldn't think of an alternative but I'd be happy to simplify this.

Reply:
I'm not sure I understand. You are saying that we used to work with `uint256` (instead of `mathint`) ... and that discarded cases where an overflow would happen.

If that is what we are trying to replicate, why are  we not doing
```require balance <= max_uint256;```
or something like that.

Here the comparison is between `balance` and `sumOfBalance`. Its a way stronger assumption than "no overflow"

Reply:
Here's [a run](https://prover.certora.com/output/6876/316c194b9cd448b4a56a8e5238434087?anonymousKey=13b47c009f652fe97457fa50466df049f7807fea) without the new Sload hook added.

Many rules fail. I'll take the example of the `transfer` run. The initial state is the following:

![Captura de pantalla 2023-09-08 a la(s) 9 27 59](https://github.com/OpenZeppelin/openzeppelin-contracts/assets/33379285/166c81a4-e430-486a-bd83-9ed9d2003132)

And the example is doing a `ERC20PermitHarness.transfer(to=0x2712, value=15)` where sender is `0x4c8`.
Because `0x2712` has `0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff6` balance, when adding `15`, the result is a balance of 5 because the uint256 was overflown.

However, the rule is checking that:

```
assert to_mathint(balanceOf(recipient)) == (recipientBalanceBefore + (holder == recipient ? 0 : amount));
```

where `recipientBalanceBefore` is `0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff6`.

Consider that since CVL2, [mathematical operations return a `mathint`](https://docs.certora.com/en/latest/docs/cvl/cvl2/changes.html?highlight=bytes32#mathematical-operations-return-mathint), therefore the `recipientBalanceBefore + amount` will result in `0x10000000000000000000000000000000000000000000000000000000000000005`.

When comparing, it turns out that `0x10000000000000000000000000000000000000000000000000000000000000005 != 5`.

The counterexample seems valid, but I couldn't find why it was not happening in CVL1

> You are saying that we used to work with uint256 (instead of mathint) ... and that discarded cases where an overflow would happen.

This is currently my strongest hypothesis but not 100% sure tbh. The reason I think so is because the casting changes added in CVL2 although they're not explicit about the previous behavior.

Also, any attempt to add an invariant that avoid this will result in a similar result. I tried a few alternatives with invariants but didn't work. Counterexamples were similar to the run I just provided

Reply:
Clearly, this counter example is the result of the initial state missing some invariant. 

We (devs) would expect the initial state to follow

- `_balances[0x4c8] <= _totalSupply`
- `_balances[0x2712] <= _totalSupply`
- `_balances[0x4c8] + _balances[0x2712] <= _totalSupply` (that last part is more problematic ...)

because there are invariants that we have in mind.

My point is, if these invariants are necessary for the rules to pass, we need to
1. express them as invariant
2. prove them
3. do a `requireInvariant` in the rules that need that assumption (and yes, there will be many).